src/analysis/inner_product_space/rayleigh.lean

```diff
@@ -0,0 +1,278 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth, Frédéric Dupuis
+-/
+import analysis.inner_product_space.calculus
+import analysis.inner_product_space.dual
+import analysis.calculus.lagrange_multipliers
+import linear_algebra.eigenspace
+
+/-!
+# The Rayleigh quotient
+
+The Rayleigh quotient of a self-adjoint operator `T` on an inner product space `E` is the function
+`λ x, ⟪T x, x⟫ / ∥x∥ ^ 2`.
+
+The main results of this file are `is_self_adjoint.has_eigenvector_of_is_max_on` and
+`is_self_adjoint.has_eigenvector_of_is_min_on`, which state that if `E` is complete, and if the
+Rayleigh quotient attains its global maximum/minimum over some sphere at the point `x₀`, then `x₀`
+is an eigenvector of `T`, and the `supr`/`infi` of `λ x, ⟪T x, x⟫ / ∥x∥ ^ 2` is the corresponding
+eigenvalue.
+
+The corollaries `is_self_adjoint.has_eigenvalue_supr_of_finite_dimensional` and
+`is_self_adjoint.has_eigenvalue_supr_of_finite_dimensional` state that if `E` is finite-dimensional
+and nontrivial, then `T` has some (nonzero) eigenvectors with eigenvalue the `supr`/`infi` of
+`λ x, ⟪T x, x⟫ / ∥x∥ ^ 2`.
+
+## TODO
+
+A slightly more elaborate corollary is that if `E` is complete and `T` is a compact operator, then
+`T` has some (nonzero) eigenvector with eigenvalue either `⨆ x, ⟪T x, x⟫ / ∥x∥ ^ 2` or
+`⨅ x, ⟪T x, x⟫ / ∥x∥ ^ 2` (not necessarily both).
+
+-/
+
+variables {𝕜 : Type*} [is_R_or_C 𝕜]
+variables {E : Type*} [inner_product_space 𝕜 E]
+
+local notation `⟪`x`, `y`⟫` := @inner 𝕜 _ _ x y
+open_locale nnreal
+
+open module.End metric
+
+namespace continuous_linear_map
+variables (T : E →L[𝕜] E)
+
+lemma rayleigh_smul (x : E) {c : 𝕜} (hc : c ≠ 0) :
+  T.re_apply_inner_self (c • x) / ∥(c • x:E)∥ ^ 2 = T.re_apply_inner_self x / ∥(x:E)∥ ^ 2 :=
+begin
+  by_cases hx : x = 0,
+  { simp [hx] },
+  have : ∥c∥ ≠ 0 := by simp [hc],
+  have : ∥x∥ ≠ 0 := by simp [hx],
+  field_simp [norm_smul, T.re_apply_inner_self_smul],
+  ring
+end
+
+lemma image_rayleigh_eq_image_rayleigh_sphere {r : ℝ} (hr : 0 < r) :
+  (λ x : E, T.re_apply_inner_self x / ∥(x:E)∥ ^ 2) '' {0}ᶜ
+  = (λ x : E, T.re_apply_inner_self x / ∥(x:E)∥ ^ 2) '' (sphere 0 r) :=
+begin
+  ext x,
+  split,
+  { rintros ⟨a, (ha : a ≠ 0), haT⟩,
+    have : ∥a∥ ≠ 0 := by simp [ha],
+    let c : 𝕜 := ↑∥a∥⁻¹ * r,
+    have : c ≠ 0 := by simp [c, ha, hr.ne'],
+    refine ⟨c • a, _, _⟩,
+    { field_simp [norm_smul, is_R_or_C.norm_eq_abs, abs_of_nonneg hr.le] },
+    convert haT using 1,
```

Comment:
Might this be better as `suffices`?

Reply:
The whole trick became unnecessary after I introduced `rayleigh_quotient` to refer to the function concisely; now it's
```lean
    { rw T.rayleigh_smul a this,
      exact haT } },
```